### PR TITLE
Hashap key generic interfaces

### PR DESCRIPTION
**Pull request by chuckyvt**
This is update to the hashmap capabilities to include generic key interfaces for most of the type-bound procedures that use key type.  I believe this will be a nice step forward for the user interface, reducing number of lines of code to use the hashamaps, and makes it closer to the style of a dictionary in Python.   
 
For example, currently to map an entry:  
```
call set( key, [1] )
call map % map_entry( key, other, conflict )
```

This simplifies to 
`call map % map_entry( [1], other, conflict )`

This is the initial submit with src and testing updated for review.  Will work on specs and adding examples next.

### CODE REVIEW

**Comment by chuckyvt** [comment]
I believe this is ready for review.  There was a single CI failure that I don't understand, but doesn't appear to be related to the code updates?

**Comment by chuckyvt** [comment]
Also worth mentioning that some of the hashmap examples were somewhat outdated, assuming the hashmap capabilities have matured since the examples were written.   In one example there was a routine written, key_to_char, to convert a key back to a character string.  Since the get function does this exact task, the example was updated to use the get function instead of using a custom function.

**Comment by chuckyvt** [comment]
They are unrelated.   This PR only affects the 'key' field.  No changes to 'other'.

File: example/hashmaps/example_hashmaps_get_other_data.f90 [hunk 1]
```diff
@@ -8,15 +8,23 @@ program example_get_other_data
...
+  class(*), allocatable       :: data
+  integer(int8), allocatable  :: key_array(:)
+  integer                     :: int_scalar
+  
+  ! Initialize hashmap
   call map%init(fnv_1_hasher)
-  call set(key, [0_int8, 1_int8, 2_int8, 3_int8, 4_int8])
-  call set(other, data)
+  
+  ! Hashmap fuctions are setup to store scalar value types (other).  Use a dervied
```
Comment (jvdp1) [inline example/hashmaps/example_hashmaps_get_other_data.f90:21]:
```suggestion
  ! Hashmap functions are setup to store scalar value types (other).  Use a dervied
```

File: example/hashmaps/example_hashmaps_map_entry.f90 [hunk 2]
```diff
@@ -1,17 +1,43 @@
...
+  int_scalar = 1
+  call map%map_entry( [int_scalar], other, conflict)
+  print *, 'CONFLICT = ', conflict
+  
+  ! Using map_entry character interface
+  call map%map_entry( 'key_string', other, conflict)
+  print *, 'CONFLICT = ', conflict
+  
+  ! Transfer unsupported key types to int8 arrays.
+  call map%map_entry( TRANSFER( [1_int64, 2_int64, 3_int64], [0_int8] ), other, conflict)
```
Comment (jvdp1) [inline example/hashmaps/example_hashmaps_map_entry.f90:37]:
```suggestion
  call map%map_entry( transfer( [1_int64, 2_int64, 3_int64], [0_int8] ), other, conflict)
```

File: example/hashmaps/example_hashmaps_remove.f90 [hunk 3]
```diff
@@ -8,12 +8,39 @@ program example_remove
...
+  ! Hashmap will dynamically increase size if needed.
   call map%init(fnv_1_hasher, slots_bits=10)
-  call set(key, [5_int8, 7_int8, 4_int8, 13_int8])
-  call set(other, dummy)
+  
+  ! Initialize other type with data to store.
+  call set(other, 4.0)
+  
+  ! Explicitly set key type using set function
+  call set(key, [1, 2, 3] )
```
Comment (jvdp1) [inline example/hashmaps/example_hashmaps_remove.f90:21]:
```suggestion
  call set(key, [1, 2, 3])
```

File: example/hashmaps/example_hashmaps_set_other_data.f90 [hunk 4]
```diff
@@ -8,15 +8,17 @@ program example_set_other_data
...
   type(other_type)    :: other
-  class(*), allocatable :: dummy
+    
+  ! Initialize hashmap with 2^10 slots.
+  ! Hashmap will dynamically increase size if needed.
   call map%init(fnv_1_hasher, slots_bits=10)
-  allocate (dummy, source='A value')
-  call set(key, [5_int8, 7_int8, 4_int8, 13_int8])
-  call set(other, dummy)
+
```
Comment (jvdp1) [inline example/hashmaps/example_hashmaps_set_other_data.f90:15]:
```suggestion

```

File: src/stdlib_hashmaps.f90 [hunk 5]
```diff
@@ -88,23 +105,53 @@ module stdlib_hashmaps
 !! Hash function
 
     contains
-
+
```
Comment (jvdp1) [inline src/stdlib_hashmaps.f90:108]:
```suggestion

```

File: src/stdlib_hashmaps.f90 [hunk 6]
```diff
@@ -770,72 +818,411 @@ end function total_open_depth
...
 
-contains
+    
+    contains
+    
+    
+        subroutine int8_get_other_data( map, value, other, exists )
+!! Version: Experimental
+!!
+!!! Int8 key generic interface for get_other_data function
```
Comment (jvdp1) [inline src/stdlib_hashmaps.f90:828]:
```suggestion
!! Int8 key generic interface for get_other_data function
```

File: test/hashmaps/test_maps.fypp [hunk 7]
```diff
@@ -24,11 +24,13 @@ module test_stdlib_chaining_maps
     integer, parameter        :: test_size = rand_size*4
     integer, parameter        :: test_16 = 2**4
     integer, parameter        :: test_256 = 2**8
-    ! key_type = 2 to support int8 and int32 key types tested.  Can be 
+    ! key_type = 5 to support int8 and int32 key types tested.  Can be 
     ! increased to generate additional unique int8 vectors additional key types.   
-    integer, parameter        :: key_types = 2
+    integer, parameter        :: key_types = 5
+    character(len=16)         :: char_type
```
Comment (jvdp1) [inline test/hashmaps/test_maps.fypp:30]: The suffix `_type` is usually reserved/used for DT in `stdlib`. Could `char_size` or `char_bytes` be an alternative?

**Review by jvdp1 — APPROVED** [review]
Thank you @chuckyvt . Overall LGTM. IMO it could be merged, pending minor changes.

File: test/hashmaps/test_maps.fypp (hunk 7 above)
Comment (chuckyvt) [reply test/hashmaps/test_maps.fypp]: This has been changed to char_size

**Comment by chuckyvt** [comment]
Thanks for the detailed review.  I believe all requested changes have been included.

**Comment by jvdp1** [comment]
@chuckyvt : Could you rebase this PR, please (to check if it solves the CI issue)?

**Comment by jvdp1** [comment]
Thank you @chuckyvt . I will wait a couple of days before merging, in case of additional comments.

**Pull request merged**